src/charm.py

```diff
@@ -0,0 +1,257 @@
+        self.unit.set_ports(8083)
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        logger.debug("_on_collect_status")
+        event.add_status(ops.ActiveStatus())
+        _, status = self._get_library_write_behaviour()
+        event.add_status(status)
+
+    def _on_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
+        """Define and start the workload using the Pebble API."""
+        container = event.workload
+        container.add_layer(SERVICE_NAME, {**self.get_pebble_layer()}, combine=True)
+        container.replan()
+
+    def _on_storage_attached(self, event: ops.StorageAttachedEvent) -> None:
+        self._push_library_to_storage()
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Don't do anything! User can run library-write action if needed.
+
+        Bad config values are handled in _on_collect_status
+        """
+        pass
+
+    def _on_library_write(self, event: ops.ActionEvent) -> None:
+        library_write_behaviour, status = self._get_library_write_behaviour()
+        if library_write_behaviour is None:
+            event.fail(status.message)
+            return
+        self._push_library_to_storage()
+        event.set_results({LIBRARY_WRITE_CONFIG: library_write_behaviour})
+
+    def _on_library_info(self, event: ops.ActionEvent) -> None:
+        container = self.framework.model.unit.containers[CONTAINER_NAME]
+        format = cast(str, event.params[LIBRARY_INFO_FORMAT_PARAM])
+        match format:
+            case 'tree':
+                try:
+                    container.exec(['which', 'tree']).wait()
+                except ops.pebble.ExecError:
+                    logger.debug('_on_library_info: tree: installing dependencies')
+                    container.exec(['apt', 'update']).wait()
+                    container.exec(['apt', 'install', 'tree', '-y']).wait()
+                logger.debug('_on_library_info: tree: executing')
+                process = container.exec(
+                    ['tree'],
+                    working_dir='/books/',
+                    stdout=cast(typing.BinaryIO, (stdout := CaptureStdOut())),
+                )
+                process.wait()
+                event.set_results({'tree': '\n'.join(stdout.lines)})
+            case 'ls-1':
+                logger.debug('_on_library_info: tree: executing')
+                process = container.exec(
+                    ['ls', '-1'],
+                    working_dir='/books/',
+                    stdout=cast(typing.BinaryIO, (stdout := CaptureStdOut())),
+                )
+                process.wait()
+                event.set_results({'ls-1': '\n'.join(stdout.lines)})
+            case _:
+                msg = (
+                    f'Invalid value {format} for {LIBRARY_INFO_FORMAT_PARAM} parameter'
+                    f' of {LIBRARY_INFO_ACTION} action.'
+                )
+                logger.error(f'_on_library_info: {format}: {msg}')
```

Comment:
You probably don't need `{format}` in here since it's inside of `msg`.

Also, it's your charm so ignore this if you like, but you shouldn't really use f-strings in logging calls. If it's written like:

```suggestion
                logger.error('_on_library_info: %s', msg)
```

Then logging handlers can be smarter in what they do, like aggregate errors based on the raw message, or skip evaluating the args (which wouldn't do anything with this plain string, of course).